src/cargo/core/compiler/output_sbom.rs

```diff
@@ -0,0 +1,239 @@
+pub struct SbomFormatVersion<const V: u32>;
+
+impl<const V: u32> Serialize for SbomFormatVersion<V> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_u32(V)
+    }
+}
+
+/// A package dependency
+#[derive(Serialize, Clone, Debug)]
+struct SbomDependency {
+    name: String,
+    package_id: PackageId,
+    version: String,
+    features: Vec<String>,
+}
+
+impl From<&UnitDep> for SbomDependency {
+    fn from(dep: &UnitDep) -> Self {
+        let package_id = dep.unit.pkg.package_id();
+        let name = package_id.name().to_string();
+        let version = package_id.version().to_string();
+        let features = dep
+            .unit
+            .features
+            .iter()
+            .map(|f| f.to_string())
+            .collect_vec();
+
+        Self {
+            name,
+            package_id,
+            version,
+            features,
+        }
+    }
+}
+
+#[derive(Serialize, Clone, Debug)]
+struct SbomPackage {
+    package_id: PackageId,
+    package: String,
+    version: String,
+    features: Vec<String>,
+    build_type: SbomBuildType,
+    extern_crate_name: String,
+    dependencies: Vec<SbomDependency>,
+}
+
+impl SbomPackage {
+    pub fn new(
+        dep: &UnitDep,
+        dependencies: Vec<SbomDependency>,
+        build_type: SbomBuildType,
+    ) -> Self {
+        let package_id = dep.unit.pkg.package_id();
+        let package = package_id.name().to_string();
+        let version = package_id.version().to_string();
+        let features = dep
+            .unit
+            .features
+            .iter()
+            .map(|f| f.to_string())
+            .collect_vec();
+
+        Self {
+            package_id,
+            package,
+            version,
+            features,
+            build_type,
+            extern_crate_name: dep.extern_crate_name.to_string(),
+            dependencies,
+        }
+    }
+}
+
+#[derive(Serialize)]
+struct SbomTarget {
+    kind: TargetKind,
+    crate_type: Option<CrateType>,
+    name: String,
+    edition: String,
+}
+
+impl From<&Target> for SbomTarget {
+    fn from(target: &Target) -> Self {
+        SbomTarget {
+            kind: target.kind().clone(),
+            crate_type: target.kind().rustc_crate_types().first().cloned(),
+            name: target.name().to_string(),
+            edition: target.edition().to_string(),
+        }
+    }
+}
+
+#[derive(Serialize, Clone)]
+struct SbomRustc {
+    version: String,
+    wrapper: Option<PathBuf>,
+    commit_hash: Option<String>,
+    host: String,
+}
+
+impl From<&Rustc> for SbomRustc {
+    fn from(rustc: &Rustc) -> Self {
+        Self {
+            version: rustc.version.to_string(),
+            wrapper: rustc.wrapper.clone(),
+            commit_hash: rustc.commit_hash.clone(),
+            host: rustc.host.to_string(),
+        }
+    }
+}
+
+#[derive(Serialize)]
+struct Sbom {
```

Comment:
Do we need a dependencies field for this top-level `Sbom`?

(Just a question. I don't really know if other SBOM formats need it to recover the dependency graph)

Reply:
Ideally, yes. Copying my comment from the RFC:

> Note that there are two ways of looking at dependencies: what each package needs, and the final resolved graph.
> 
> For example, if one package depends on `rand` with `features = ["std", "getrandom"]`, and another with `features = ["std", "simd_support"]`, the final resolved features will be `["std", "getrandom", "simd_support"]`. Depending on the use case you may need either or both representations (direct package dependencies and the resolved graph).
> 
> `cargo metadata` [exposes both](https://doc.rust-lang.org/cargo/commands/cargo-metadata.html) (under "packages" and "resolve" fields), but inaccurately: 
>  - https://github.com/rust-lang/cargo/issues/7754
>  - Under resolver v2, it [conflates the normal and build-dependency trees](https://rust-lang.github.io/rfcs/2957-cargo-features2.html#cargo-metadata)
> 
> I think it would be best for the SBOM to also expose both, accurately this time.

So what I would like to see is *two* resolved dependency trees: one for normal dependencies and one for build dependencies, matching the way feature resolver v2 works.